ICLM-15: Added FHIR related dependencies and insurance claim diagnosis mapping

dborowiecki opened the pull request:
### Summary
FHIR dependencies to module
Added methods for transforming InsuranceClaimDiagnosis to DiagnosisComponent and vice versa
Added methods responsible for finding a diagnosis conncected with InsuranceClaim

Ticket: [ICLM-15](https://issues.openmrs.org/projects/ICLM/issues/ICLM-15?filter=allopenissues)

dborowiecki committed:
Added FHIR related dependencies and insurance claim diagnosis mapping

dborowiecki committed:
added trusty dist to travis

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/InsuranceClaimDao.java:
```diff
@@ -1,8 +1,16 @@
 
 public interface InsuranceClaimDao extends OpenmrsDataDAO<InsuranceClaim> {
 
+    List<VisitType> findVisitTypeByName(String visitTypeName);
```
Probably it should be rarther in VisitType dao

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImpl.java:
```diff
@@ -1,14 +1,59 @@
 	}
+
+
+	private DbSession getSession() {
```
private methods should be at the bottom of page

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImpl.java:
```diff
@@ -1,14 +1,59 @@
+	 * @param ic - InsuranceClaim object
+	 */
+	@Override
+	public List<InsuranceClaimDiagnosis> findInsuranceClaimDiagnosis(InsuranceClaim ic) {
```
i think there is no need to send InsuranceClaim object, send only id

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImpl.java:
```diff
@@ -1,14 +1,59 @@
+	 * Finds all VisitTypes with name given name
+	 *
+	 * @param visitTypeName - Name of visit type
+	 */
```
lack of override annotation

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImpl.java:
```diff
@@ -1,14 +1,59 @@
+		crit.add(Restrictions.eq("claim.id", ic.getId()));
+		List<InsuranceClaimDiagnosis> result = crit.list();
+
+		return result;
```
why you retrieve InsuranceClaimDiagnosis in InsuranceClaim Dao ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImpl.java:
```diff
@@ -1,14 +1,59 @@
+	 */
+	@Override
+	public List<InsuranceClaimDiagnosis> findInsuranceClaimDiagnosis(InsuranceClaim ic) {
+		Criteria crit = getSession().createCriteria(InsuranceClaimDiagnosis.class, "d");
```
please omit using short, not readable paramatere name like 'd'

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/DiagnosisHelper.java:
```diff
@@ -0,0 +1,14 @@
+import org.openmrs.module.insuranceclaims.api.model.InsuranceClaim;
+import org.openmrs.module.insuranceclaims.api.model.InsuranceClaimDiagnosis;
+
+
```
remove empty line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/DiagnosisHelper.java:
```diff
@@ -0,0 +1,14 @@
+    void setConcept(InsuranceClaimDiagnosis diagnosis, Concept concept);
+
+    void setInsuranceClaim(InsuranceClaimDiagnosis diagnosis, InsuranceClaim claim);
+
```
remove empty line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/DiagnosisHelper.java:
```diff
@@ -0,0 +1,14 @@
+import org.openmrs.module.insuranceclaims.api.model.InsuranceClaimDiagnosis;
+
+
+public interface DiagnosisHelper {
```
this class does not make sense (with impl)

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/InsuranceClaimHelper.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+public interface InsuranceClaimHelper {
```
this class does not make sense - use dao instead

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/InsuranceClaimHelper.java:
```diff
@@ -0,0 +1,13 @@
+
+    List<InsuranceClaimDiagnosis> getInsuranceClaimDiagnosis(InsuranceClaim claim);
+
+
```
please remebmer about removing empty lines

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/impl/DiagnosisHelperImpl.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.stereotype.Component;
+
+@Component(value = "insuranceclaims.DiagnosisHelper")
+public class DiagnosisHelperImpl implements DiagnosisHelper {
```
remove class

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/impl/DiagnosisHelperImpl.java:
```diff
@@ -0,0 +1,27 @@
+public class DiagnosisHelperImpl implements DiagnosisHelper {
+
+    @Autowired
+    InsuranceClaimDiagnosisDao dao;
```
fields should always have moderator - private here

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/helper/impl/InsuranceClaimHelperImpl.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+@Component(value = "insuranceclaims.InsuranceClaimHelper")
+public class InsuranceClaimHelperImpl implements InsuranceClaimHelper {
```
remove this class

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/util/ContextUtil.java:
```diff
@@ -0,0 +1,21 @@
+import org.openmrs.module.insuranceclaims.api.helper.DiagnosisHelper;
+import org.openmrs.module.insuranceclaims.api.helper.InsuranceClaimHelper;
+
+public class ContextUtil {
```
should be final, however i do not see sense of this method. Just use autowired in different approach

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/util/FHIRClaimDiagnosisUtil.java:
```diff
@@ -0,0 +1,59 @@
+
+public class FHIRClaimDiagnosisUtil {
+
+    public static Claim.DiagnosisComponent createClaimDiagnosisComponent(InsuranceClaimDiagnosis omrsClaimDiagnosis) {
```
import Diagnosis Component

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/util/FHIRClaimDiagnosisUtil.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class FHIRClaimDiagnosisUtil {
```
should be final with private constructor. Why you use so many util classes ?

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImplTest.java:
```diff
@@ -1,19 +1,28 @@
 package org.openmrs.module.insuranceclaims.api.dao.impl;
 
+
```
please remove empty line

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImplTest.java:
```diff
@@ -22,6 +31,16 @@
+	@Autowired
+	private InsuranceClaimDiagnosisDao diagnosisDao;
+
+	private InsuranceClaim insuranceClaim;
```
i do not see usage of it

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/helpers/impl/TestDiagnosisHelper.java:
```diff
@@ -0,0 +1,71 @@
+    private InsuranceClaimDiagnosisDao insuranceClaimDiagnosisDao;
+
+    @Autowired
+    DiagnosisHelper diagnosisHelper;
```
please use moderator

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/helpers/impl/TestInsuranceClaimHelper.java:
```diff
@@ -0,0 +1,73 @@
+    InsuranceClaimHelper insuranceClaimHelper;
+
+    InsuranceClaim insuranceClaim;
+    InsuranceClaimDiagnosis insuranceClaimDiagnosis;
```
use moderators

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/testutils/TestConstants.java:
```diff
@@ -1,4 +1,4 @@
-package org.openmrs.module.insuranceclaims.api.util;
+package org.openmrs.module.insuranceclaims.api.testutils;
```
why did you rename this package ?

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/util/TestContextUtil.java:
```diff
@@ -0,0 +1,28 @@
+import org.openmrs.module.insuranceclaims.util.ContextUtil;
+import org.openmrs.test.BaseModuleContextSensitiveTest;
+
+public class TestContextUtil extends BaseModuleContextSensitiveTest {
```
ContextUtilTest

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/util/TestContextUtil.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Test
+    public void testGetInsuranceClaimHelperInstance() {
```
test naming convention is strange, please use openmrs convention

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/util/TestFHIRClaimDiagnosisUtil.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.List;
+import java.util.logging.Logger;
+
+
```
remove empty line

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/util/TestFHIRClaimDiagnosisUtil.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.logging.Logger;
+
+
+public class TestFHIRClaimDiagnosisUtil extends BaseModuleContextSensitiveTest {
```
Test at the end of class name

dborowiecki committed:
ICLM-15 Code refactor, remove diagnosis util and create diagnosis test

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/InsuranceClaimDiagnosisDao.java:
```diff
@@ -3,6 +3,9 @@
+
 public interface InsuranceClaimDiagnosisDao extends OpenmrsDataDAO<InsuranceClaimDiagnosis> {
 
+    List<InsuranceClaimDiagnosis> findInsuranceClaimDiagnosis(Integer insuranceClaimId);
```
can we use here int ? If possible, we should always use primitivies

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDiagnosisDaoImpl.java:
```diff
@@ -1,15 +1,44 @@
+		return result;
+	}
+
+	private DbSession getSession() {
```
getCurrentSession

kmadej commented on pom.xml:
```diff
@@ -61,6 +67,52 @@
+			<dependency>
+				<groupId>ca.uhn.hapi.fhir</groupId>
+				<artifactId>hapi-fhir-server</artifactId>
+				<version>${hapiFhirStructuresDstuVersion}</version>
```
be sure that we use the proper version - similar to OpenIMIS

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+    private InsuranceClaimDiagnosisDao diagnosisDao;
+
+    public FHIRClaimDiagnosisServiceImpl() {
```
do we need this default constructor ?

dborowiecki replied:
The default constructor was needed because in the absence of it exception occurs when bean associated with this class is created.

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.LinkedList;
+import java.util.List;
+
+@Transactional
```
why you add transactional ? Do we need this ?

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDaoImplTest.java:
```diff
@@ -59,4 +59,5 @@ private InsuranceClaim createTestInstance() {
 				.getPatientIdentifierType(TestConstants.TEST_IDENTIFIER_TYPE_ID);
 		return InsuranceClaimMother.createTestInstance(location, provider, visitType, identifierType);
 	}
+
```
remove empty line

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/service/FHIRClaimDiagnosisServiceTest.java:
```diff
@@ -0,0 +1,141 @@
+        Concept concept = Context.getConceptService().getConceptByUuid("160148AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
+        return InsuranceClaimDiagnosisMother.createTestInstance(concept, this.testInsuranceClaim);
+    }
+
```
remove empty line

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/service/FHIRClaimDiagnosisServiceTest.java:
```diff
@@ -0,0 +1,141 @@
+            executeDataSet(conceptPath);
+        } catch (Exception e) {
+            Logger log = Logger.getGlobal();
+            log.warning("Problem with loading concept from malaria-concept.xml: \n"
```
it is only test, do not use it. Just throw exception

kmadej commented on pom.xml:
```diff
@@ -30,13 +30,19 @@
 
 	<properties>
 		<javaVersion>1.8</javaVersion>
+		<openMRSVersion>2.4.0-SNAPSHOT</openMRSVersion>
```
why do you use snapshot version ?

dborowiecki committed:
ICLM-15: Exclude tests from checkstyle and refactor

dborowiecki committed:
ICLM-15: Added mapping for Claim

dborowiecki committed:
ICLM-15: Changes in InsuranceClaimUtil layout

dborowiecki commented:
@kmadej I've made the changes you recommended in the comments and added functionalities responsible for mapping the InsuranceClaim to FHIR Claim

dborowiecki committed:
ICLM-15: Added mapping for ClaimRepsonse

dborowiecki committed:
ICLM-15: InsuranceClaim uses claimCode for mapping fhir claim id instead of uuid

dborowiecki committed:
ICLM-15: Code cleanup

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/dao/impl/InsuranceClaimDiagnosisDaoImpl.java:
```diff
@@ -1,15 +1,44 @@
+
+		crit.add(Restrictions.eq("claim.id", insuranceClaimId));
+		List<InsuranceClaimDiagnosis> result = crit.list();
+		return result;
```
return crit.list()

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/model/InsuranceClaimStatus.java:
```diff
@@ -4,5 +4,9 @@
+	REJECTED, ENTERED, CHECKED, PROCESSED, VALUATED;
+
+	public int getValue() {
+		return ordinal() + 1;
```
needs explanation

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+    public FHIRClaimDiagnosisServiceImpl() {
+    }
+
+    public FHIRClaimDiagnosisServiceImpl(InsuranceClaimDiagnosisDao diagnosisDao) {
```
I see that in xml you autowired in using property - setter and constructor are redundant in this case

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+    public FHIRClaimDiagnosisServiceImpl(InsuranceClaimDiagnosisDao diagnosisDao) {
+        this.diagnosisDao = diagnosisDao;
+    }
+
```
add override

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+        return allDiagnosisComponents;
+    }
+
+    @Transactional
```
to use transactional in openmrs spring beans you need to wrap it using TransactionalProxyBean

kmadej replied:
if you use domain class in method it should be wrapped in transactional

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+    public List<Claim.DiagnosisComponent> createClaimDiagnosisComponent(
+            List<InsuranceClaimDiagnosis> omrsClaimDiagnosis) {
+        List<Claim.DiagnosisComponent> allDiagnosisComponents = new LinkedList<>();
+        diagnosisDao.hashCode();
```
what is going on ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+    public List<Claim.DiagnosisComponent> createClaimDiagnosisComponent(
+            List<InsuranceClaimDiagnosis> omrsClaimDiagnosis) {
+        List<Claim.DiagnosisComponent> allDiagnosisComponents = new LinkedList<>();
```
why you use linkedlist ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+        try {
+            diagnosis.setConcept(FHIRUtils.getConceptFromCode(claimDiagnosis.getDiagnosisCodeableConcept(), errors));
+        } catch (Exception e) {
```
please do not catch generic Exception

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+        this.diagnosisDao = diagnosisDao;
+    }
+
+    public Claim.DiagnosisComponent createClaimDiagnosisComponent(InsuranceClaimDiagnosis omrsClaimDiagnosis) {
```
change create to generate

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,100 @@
+                .getText();
+        try {
+            return InsuranceClaimStatus.valueOf(codeString);
+        } catch (ArrayIndexOutOfBoundsException e) {
```
you should not catch runtime exceptions

kmadej replied:
again

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,100 @@
+
+    public static List<Reference> buildCommunicationRequestReference(InsuranceClaim omrsClaim) {
+        Reference reference = new Reference();
+        String stringReference = "CommunicationRequest/" + omrsClaim.getUuid();
```
please use fhir constanst

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,100 @@
+            reasonCoding.setCode(rejectionReason);
+        }
+        else {
+            reasonCoding.setCode("0");
```
"0" define it as constant

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,100 @@
+        }
+        codeableConcept.setCoding(Collections.singletonList(reasonCoding));
+        error.setCode(codeableConcept);
+        return Collections.singletonList(error);
```
please remember that singletonList is not modifeable

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,100 @@
+
+    public static String getClaimResponseErrorCode(ClaimResponse claimResponse) {
+        ClaimResponse.ErrorComponent ec = claimResponse.getErrorFirstRep();
+        return ec.getCode().getCoding().get(0).getCode();
```
this line it should be more readable

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,100 @@
+        return ec.getCode().getCoding().get(0).getCode();
+    }
+
+    public static ClaimResponse.PaymentComponent createPaymentComponent(InsuranceClaim omrsClaim) {
```
final class should have private constructor at the bottom

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,85 @@
+    public static String getIdentifierValueByCode(Claim claim, String code, List<String> errors) {
+        try {
+            return getClaimIdentifierValueBySystemCode(claim, code);
+        } catch (Exception e) {
```
do not catch Exception

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    private static String validatedIdentifier(List<Identifier> claimCodeIdentifier, String code) throws FHIRException {
+        if (claimCodeIdentifier.size() != 1) {
```
use utils methods

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,85 @@
+            throw new FHIRException("Could not found unique identifier for system code "
+                    + code);
+        }
+        return claimCodeIdentifier.get(0).getValue();
```
not readable enough

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    private static List<Identifier> getIdentifierBySystemCode(List<Identifier> identifierList, String code) {
+        List<Identifier> result = new LinkedList<>();
```
hashlist

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,85 @@
+    private static List<Identifier> getIdentifierBySystemCode(List<Identifier> identifierList, String code) {
+        List<Identifier> result = new LinkedList<>();
+
+        for (Identifier i : identifierList) {
```
i change to identifier

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimConstants.java:
```diff
@@ -0,0 +1,7 @@
+public final class InsuranceClaimConstants {
+    public static final String MEDICAL_RECORD_NUMBER = "MR";
+    public static final String ACCESSION_ID = "ACSN";
+    public static final String HL7_VALUESET_SYSTEM = "https://hl7.org/fhir/valueset-identifier-type.html";
```
private constructor

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+import static org.openmrs.module.insuranceclaims.api.service.fhir.util.SpecialComponentUtil.getSpecialConditionComponentFromCategory;
+
+public final class InsuranceClaimUtil {
+    public static List<Claim.DiagnosisComponent> getClaimDiagnosis(InsuranceClaim omrsClaim) {
```
it should not be in util class

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+                .createClaimDiagnosisComponent(omrsClaim);
+    }
+
+    public static Reference buildLocationReference(Location location) {
```
maybe move it to LocationUtil

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+            errors.add("More than one matching visit type was found: %");
+            return null;
+        }
+        return visitType.get(0);
```
why get(0) ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+    public static VisitType getClaimVisitType(Claim claim, List<String> errors) {
+        String visitTypeName = claim.getType().getText();
+        List<VisitType> visitType = getVisitTypeByName(visitTypeName);
+        if (visitType.size() > 1) {
```
why .size() > 1, what if visitType will be null ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+
+    public static Patient getClaimPatient(Claim claim, List<String> errors) {
+        String patientUuid = FHIRUtils.getObjectUuidByReference(claim.getPatient());
+        Patient patient = Context.getPatientService().getPatientByUuid(patientUuid);
```
not good method in util class

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+
+    public static Provider getClaimProvider(Claim claim, List<String> errors) {
+        String providerUUid = FHIRUtils.getObjectUuidByReference(claim.getProvider());
+        Provider provider = Context.getProviderService().getProviderByUuid(providerUUid);
```
the same

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+
+    public static Location getClaimLocation(Claim claim, List<String> errors) {
+        String locationUuid = FHIRUtils.getObjectUuidByReference(claim.getFacility());
+        Location location = Context.getLocationService().getLocationByUuid(locationUuid);
```
the same

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,181 @@
+        if (to == null) {
+            errors.add("Date 'to' is missing");
+        }
+        claimPeriod.put("from", from);
```
"from" should be defined as constant and used also while getting information from this map

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/service/FHIRClaimDiagnosisServiceTest.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Autowired
+    @Qualifier("insuranceclaims.fhirDiagnosis")
+    private FHIRClaimDiagnosisService util;
```
why you name it util ?

kmadej commented on api/src/test/java/org/openmrs/module/insuranceclaims/api/service/FHIRClaimResponseServiceImplTest.java:
```diff
@@ -0,0 +1,119 @@
+    @Test
+    public void generateFhirClaimResponse_shouldMapInsuranceClaimToFhirClaimResponse() throws FHIRException {
+        InsuranceClaim savedInsuranceClaim = insuranceClaimDao.getByUuid(insuranceClaim.getUuid());
+        FHIRClaimResponseService insuranceClaimService = new FHIRClaimResponseServiceImpl();
```
do it similar as in above test class

dborowiecki committed:
ICLM-15: Added externalId attributes

dborowiecki committed:
ICLM-15: Merge model changes to branch

dborowiecki committed:
ICLM-15: Added mapping for claim item

dborowiecki committed:
ICLM-15: Code cleanup and refactor

dborowiecki committed:
ICLM-15: Fixed application context beans

dborowiecki committed:
ICLM-15: Merged changes from master

kmadej commented on api/pom.xml:
```diff
@@ -25,4 +25,55 @@
+            <artifactId>fhir-api</artifactId>
+        </dependency>
+
+<!--        <dependency>-->
```
why these lines are commented ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/db/impl/AttributeServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+                .collect(Collectors.toList());
+    }
+
+    public  void setSessionFactory(SessionFactory sessionFactory) {
```
do we need this ? i see you use autowired

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/db/impl/AttributeServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    public List<Location> getLocationByExternalIdAttribute(String externalId) {
```
i think it can be done more generic, all those methods

dborowiecki replied:
Although these methods look similar, there is no way to do this in a more generic way by the fact that each object returns the desired attribute in different way.

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/FHIRInsuranceClaimService.java:
```diff
@@ -0,0 +1,14 @@
+    Claim generateClaim(InsuranceClaim omrsClaim);
+
+    InsuranceClaim generateOmrsClaim(Claim claim, List<String> errors);
+
```
remove empty line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/FHIRInsuranceClaimService.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.List;
+
+public interface FHIRInsuranceClaimService {
+
```
remove empty line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+
+    @Autowired
+    private InsuranceClaimDiagnosisDao diagnosisDao;
+
```
add override

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+        return diagnosis;
+    }
+
+    public void setDiagnosisDao(InsuranceClaimDiagnosisDao diagnosisDao) {
```
do we need this ? i see you used autowired

kmadej replied:
We should remove all autowired from main code

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+
+                diagnosis.setType(diagnosisType);
+            } catch (FHIRException exception) {
+                continue;
```
i do not get what is going on here, please omit using continue in catch clause

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+        String primaryCoding = InsuranceClaimConstants.PRIMARY_DIAGNOSIS_MAPPING;
+
+        List<Coding> diagnosisCoding = diagnosis.getDiagnosisCodeableConcept().getCoding();
+        for (Coding c: diagnosisCoding) {
```
name variable coding

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+
+    private void validateDiagnosisCodingSystem(Claim.DiagnosisComponent diagnosis) {
+        //Method assigns diagnosis system (ICD-10, CIEL, etc.) from type to the concept coding if it don't have assigned system
+        if (diagnosis.getType().size() == 0) {
```
use collection utils is empty

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+        }
+
+        try {
+            String sys = diagnosis.getType().get(0).getText();
```
get(0) ?

kmadej replied:
sys ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+                }
+            }
+        } catch (FHIRException e) {
+            return;
```
do not use return in catch clause

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    private static Concept getConceptFromCode(CodeableConcept codeableConcept, List<String> errors) {
```
why static ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+        String conceptCode;
+        String system;
+        Concept concept = null;
+        List<Coding> dts = codeableConcept.getCoding();
```
dts ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+            conceptCode = coding.getCode();
+            system = coding.getSystem();
+            if (FHIRConstants.OPENMRS_URI.equals(system)) {
+                concept = Context.getConceptService().getConceptByUuid(conceptCode);
```
why not use autowired conceptservice ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+            if (FHIRConstants.OPENMRS_URI.equals(system)) {
+                concept = Context.getConceptService().getConceptByUuid(conceptCode);
+            } else {
+                String systemName = system;
```
why new field ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+                concept = Context.getConceptService().getConceptByUuid(conceptCode);
+            } else {
+                String systemName = system;
+                if (systemName != null && !systemName.isEmpty()) {
```
use StringUtils

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+                String systemName = system;
+                if (systemName != null && !systemName.isEmpty()) {
+                    List<Concept> concepts = Context.getConceptService().getConceptsByMapping(conceptCode, systemName);
+                    if (concepts.size() == 1) {
```
why we verify size == 1 ? it is not readable, maybe add new private method or change 1 to constant

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+                break;
+            }
+        }
+        if (concept == null) {
```
it can be write shorter and better

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        return insuranceClaimItems;
+    }
+
+    public void setInsuranceClaimItemDao(InsuranceClaimItemDao insuranceClaimItemDao) {
```
why you added setter ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        if (informationSequenceId == null) {
+            return null;
+        }
+        return getSpecialConditionComponentBySequenceNumber(claim, informationSequenceId);
```
do it in one line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        if (item.getInformationLinkId().isEmpty()) {
+            return null;
+        } else {
+            return item.getInformationLinkId().get(0).getValue();
```
do it in one line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    private Integer getItemComponentInformationLinkId(Claim.ItemComponent item) {
+        if (item.getInformationLinkId().isEmpty()) {
```
use collection utils

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIREligibilityRequestImpl.java:
```diff
@@ -0,0 +1,37 @@
+        request.setIdentifier(identifier);
+
+        //Patient (Should use chfid for reference)
+        Reference patient = new Reference().setReference("Patient/" + claim.getPatient().getUuid());
```
use constant for Patient path

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIREligibilityRequestImpl.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public InsurancePolicy generateEligibilityResponse(EligibilityResponse patient) {
+        return null;
```
?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRInsuranceClaimServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+package org.openmrs.module.insuranceclaims.api.service.fhir.impl;
+
+
```
remove empty line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRInsuranceClaimServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+public class FHIRInsuranceClaimServiceImpl implements FHIRInsuranceClaimService {
+
+    @Autowired
+    private AttributeService attributeService;
```
i see you use autowired however is this spring bean ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRInsuranceClaimServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+        return claim;
+    }
+
+    public InsuranceClaim generateOmrsClaim(Claim claim, List<String> errors) {
```
override

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRInsuranceClaimServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+        return omrsClaim;
+    }
+
+    public void setAttributeService(AttributeService attributeService) {
```
do we nees setter ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRInsuranceClaimServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+    private Patient getClaimPatientByExternalIdentifier(Claim claim) {
+        String patientExternalId = getIdFromReference(claim.getPatient());
+        List<Patient> patientsWithExternalId = attributeService.getPatientByExternalIdIdentifier(patientExternalId);
+        return getUnambiguousElement(patientsWithExternalId);
```
those 3 methods above look similar, can we do it more generic way ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,119 @@
+        if (hasErrors(ec)) {
+            return getCodeableConceptFirstCode(ec.getCode());
+        } else {
+            return null;
```
do it in one line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/ClaimResponseUtil.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private static boolean hasErrors(ClaimResponse.ErrorComponent errorComponent) {
+        return errorComponent.getCode().getCoding().size() != 0;
```
collection utils

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,107 @@
+        try {
+            //References are in format: Category/referenceCode
+            return reference.getReference().split("/")[1];
+        } catch (NullPointerException e) {
```
I will try to pretend that I did not see it , change ASAP

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public static <T> T getUnambiguousElement(List<T> listOfElements) {
+        if (listOfElements.isEmpty()) {
```
Collection Utils

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,107 @@
+        if (listOfElements.isEmpty()) {
+            return null;
+        }
+        Set<T> set = new HashSet<>(listOfElements);
```
set is not best name

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,107 @@
+            return null;
+        }
+        Set<T> set = new HashSet<>(listOfElements);
+        if (set.size() == 1) {
```
Collection Utils

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/IdentifierUtil.java:
```diff
@@ -0,0 +1,107 @@
+        } else {
+            throw new IdentifierNotUniqueException("Could not get unambiguous element of type "
+                    + listOfElements.get(0).getClass()
+                    + " Elements:\n" + listOfElements.toString());
```
use constant

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimConstants.java:
```diff
@@ -0,0 +1,30 @@
+    public static final String HL7_VALUESET_SYSTEM = "https://hl7.org/fhir/valueset-identifier-type.html";
+    public static final int ENUMERATION_FROM = 1;
+
+
```
remove empty line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimItemUtil.java:
```diff
@@ -0,0 +1,92 @@
+            return item.getQuantity().getValue().intValue();
+        }
+        else {
+            return null;
```
one line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimItemUtil.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private static float getConceptUnitPrice(Concept concept) {
+        return (Float) getConceptAttributeValueByTypeUuid(concept, CONCEPT_PRICE_ATTRIBUTE_UUID, null);
```
why you pass here null ?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/PractitionerUtil.java:
```diff
@@ -0,0 +1,32 @@
+
+import static org.openmrs.module.insuranceclaims.api.service.fhir.util.InsuranceClaimConstants.PROVIDER_EXTERNAL_ID_ATTRIBUTE_UUID;
+
+public final class PractitionerUtil {
```
do you need to do it as static method ? those utils above looks similar. I think it can be done with object programming

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/SpecialComponentUtil.java:
```diff
@@ -0,0 +1,62 @@
+        if (component != null) {
+            return component.getValueStringType().getValue();
+        } else {
+            return null;
```
one line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/SpecialComponentUtil.java:
```diff
@@ -0,0 +1,62 @@
+        if (requested == null) {
+            return null;
+        } else {
+            return requested.getValueStringType().getValue();
```
one line

kmadej commented on api/src/main/resources/liquibase.xml:
```diff
@@ -379,4 +379,133 @@
+        <comment>
+            Item/Service price
+        </comment>
+        <insert tableName="concept_reference_source">
```
why you inserting concept ?

dborowiecki replied:
They are added to allow reference to these specific concept values when building claims. Thanks to this, it is known in which field information about the external id of a service as well as its price are stored.

dborowiecki committed:
ICLM-15: Added mapping for items in ClaimResponse

dborowiecki committed:
ICLM-15: Code refactor

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/db/impl/AttributeServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+                .collect(Collectors.toList());
+    }
+
+    private List getOmrsObjectAttributesByAttributeTypeValue(
```
private method at the bottom

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/FHIRClaimDiagnosisService.java:
```diff
@@ -0,0 +1,22 @@
+
+    InsuranceClaimDiagnosis createOmrsClaimDiagnosis(Claim.DiagnosisComponent claimDiagnosis, List<String> errors);
+
+    void setDiagnosisDao(InsuranceClaimDiagnosisDao dao);
```
??

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/FHIRClaimItemService.java:
```diff
@@ -0,0 +1,25 @@
+
+    List<ClaimResponse.NoteComponent> generateClaimResponseNotes(InsuranceClaim claim);
+
+    void setInsuranceClaimItemDao(InsuranceClaimItemDao insuranceClaimItemDao);
```
??

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/FHIRInsuranceClaimService.java:
```diff
@@ -0,0 +1,16 @@
+
+    InsuranceClaim generateOmrsClaim(Claim claim, List<String> errors);
+
+    void setAttributeService(AttributeService attributeService);
```
??

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,170 @@
+        Concept concept = null;
+        if (FHIRConstants.OPENMRS_URI.equals(systemName)) {
+            concept = conceptService.getConceptByUuid(conceptCode);
+        } else {
```
else if in one line

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java:
```diff
@@ -0,0 +1,170 @@
+        if (FHIRConstants.OPENMRS_URI.equals(systemName)) {
+            concept = conceptService.getConceptByUuid(conceptCode);
+        } else {
+            if (isNotEmpty(systemName)) {
```
do not import static methods from external libraries - it is not readable

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,197 @@
+
+            nextItem.setSequenceLinkId(sequence);
+            nextItem.addNoteNumber(sequence);
+            sequence += 1;
```
?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,197 @@
+            nextNote.setNumber(noteNumber);
+
+            claimNotes.add(nextNote);
+            noteNumber += 1;
```
?

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimItemServiceImpl.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    private int getFirstItemNoteNumber(ClaimResponse.ItemComponent item) {
+        return item.getNoteNumber().get(0).getValue();
```
.get(0)

kmadej commented on api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/util/InsuranceClaimUtil.java:
```diff
@@ -0,0 +1,125 @@
+        Period billablePeriod = claim.getBillablePeriod();
+        Date from = billablePeriod.getStart();
+        if (from == null) {
+            errors.add("Date 'from' is missing");
```
string as static

dborowiecki committed:
ICLM-15: Remove autowired from main, remove explicit sequence incrementation, cleanup

dborowiecki committed:
ICLM-15: Remove empty lines and CollectionsUtil direct reference

dborowiecki committed:
ICLM-15: Changed FHIR version so it matches FHIR API module

dborowiecki committed:
ICLM-15: Added autojoining and autoincrementation to database tables

dborowiecki committed:
ICLM-15: Refactored beans creation

dborowiecki committed:
ICLM-15: Remove autowired from services and add new ones for db conncetion

dborowiecki committed:
ICLM-15: Removed component stereotypes from dao

dborowiecki committed:
ICLM-15: FHIR Services bug fix and adapted to changes in other services

dborowiecki committed:
ICLM-15: Added spotbugs exclude file and global api auth property

dborowiecki committed:
ICLM-15: Resolve conflicts with master

dborowiecki committed:
ICLM-15: Removed clientComponent import from app context

dborowiecki commented:
Hi @kmadej, I've added requested changes, fix some of the mappings, removed @Autowired from services and change models and database migrations so they use autoincrementation. 
Could you review it?

kmadej reviewed (approved):
Looks good to me
